### PR TITLE
Add action for creating admin user

### PR DESCRIPTION
**Pull request by arturo-seijas**
Add action for creating admin user. The action requires emails and password as parameters.
Also removes the requirement of having set up the SMTP settings for the charm to reach active status

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -453,6 +449,36 @@ def _on_database_changed(self, event: pgsql.DatabaseChangedEvent) -> None:  # ty
...
+            )
+            try:
+                process.wait_output()
+            except ExecError as ex:
+                logger.error(ex)
+                event.fail(
+                    # Parameter validation errors are printed to stdout
+                    f"Failed to create user with email {email}: {ex.stdout}"  # type: ignore
+                )
+            event.set_results({"user": f"{email}"})
```
Comment (nrobinaubertin) [inline src/charm.py:480]: Should we set this result if the action fails ?

Comment (arturo-seijas) [reply src/charm.py]: Fixed. Thanks

File: src/charm.py [hunk 2]
```diff
@@ -453,6 +449,36 @@ def _on_database_changed(self, event: pgsql.DatabaseChangedEvent) -> None:  # ty
...
+                ],
+                user="discourse",
+                working_dir=DISCOURSE_PATH,
+                environment=self._create_discourse_environment_settings(),
+            )
+            try:
+                process.wait_output()
+                event.set_results({"user": f"{email}"})
+            except ExecError as ex:
+                logger.error(ex)
```
Comment (gregory-schiano) [inline src/charm.py:476]: Perhaps logging stderr here also, as the stacktrace doesn't necessarily contains useful information in the case of a pebble exec

File: src/charm.py [hunk 3]
```diff
@@ -453,6 +449,36 @@ def _on_database_changed(self, event: pgsql.DatabaseChangedEvent) -> None:  # ty
...
+        """
+        email = event.params["email"]
+        password = event.params["password"]
+        container = self.unit.get_container("discourse")
+        if container.can_connect():
+            process = container.exec(
+                [
+                    "bash",
+                    "-c",
+                    f"./bin/bundle exec rake admin:create <<< $'{email}\n{password}\n{password}\nY'",
```
Comment (gregory-schiano) [inline src/charm.py:466]: Do you know what happens if a user with this email address already exists ?

File: src/charm.py (hunk 2 above)
Comment (arturo-seijas) [reply src/charm.py:476]: It does in this case

File: tests/unit/test_charm.py [hunk 4]
```diff
@@ -107,17 +92,9 @@ def test_config_changed_when_no_cors(self):
         assert: it will get to blocked status waiting for it.
         """
         self.add_database_relations()
+        self.harness.update_config({"cors_origin": ""})
```
Comment (merkata) [inline tests/unit/test_charm.py:95]: Is this related to the creation of the admin user?

File: tests/unit/test_charm.py [hunk 5]
```diff
@@ -130,17 +107,9 @@ def test_config_changed_when_throttle_mode_invalid(self):
         assert: it will get to blocked status waiting for a valid value to be provided.
         """
         self.add_database_relations()
+        self.harness.update_config({"throttle_level": "Scream"})
```
Comment (merkata) [inline tests/unit/test_charm.py:110]: :scream:

**Review by merkata — APPROVED** [review]
Looks good!

File: tests/unit/test_charm.py (hunk 4 above)
Comment (arturo-seijas) [reply tests/unit/test_charm.py:95]: As part of this PR we're also removing the mandatory SMTP settings as a requirement to reach active status

**Review by gregory-schiano — APPROVED** [review]
LGTM

File: tests/unit/test_charm.py [hunk 6]
```diff
@@ -399,8 +344,36 @@ def test_db_relation(self):
             "database name should be set after relation joined",
         )
 
+    def test_add_admin_user(self):
```
Comment (jdkandersson) [inline tests/unit/test_charm.py:347]: We have a standard for the test docstring, is it applicable here? https://github.com/canonical/is-charms-contributing-guide/blob/main/CONTRIBUTING.md#test-structure

File: tests/unit/test_charm.py [hunk 7]
```diff
@@ -399,8 +344,36 @@ def test_db_relation(self):
...
+        self.harness.disable_hooks()
+        self.add_database_relations()
+
+        charm: DiscourseCharm = typing.cast(DiscourseCharm, self.harness.charm)
+        with patch.object(Container, "exec", return_value=MockExecProcess()) as exec_mock:
+            self.harness.container_pebble_ready("discourse")
+
+            email = "sample@email.com"
+            password = "somepassword"  # nosec
+            event = MagicMock()
```
Comment (jdkandersson) [inline tests/unit/test_charm.py:357]: It is possible to supply a spec to Mocks using the `spec` parameter to ensure that all the interactions conform to a class or instance: https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock

File: tests/unit/test_charm.py [hunk 8]
```diff
@@ -399,8 +344,36 @@ def test_db_relation(self):
...
+                "email": email,
+                "password": password,
+            }
+            charm._on_add_admin_user_action(event)
+
+        exec_mock.assert_any_call(
+            [
+                "bash",
+                "-c",
+                "./bin/bundle exec rake admin:create <<< $'sample@email.com\nsomepassword\nsomepassword\nY'",
```
Comment (jdkandersson) [inline tests/unit/test_charm.py:368]: Would it be worth changing this to an f-string with the email and password defined earlier? E.g., `f"./bin/bundle exec rake admin:create <<< $'{email}\n{password}\n{password}\nY'"`
